source/mir/bignum/low_level_view.d

```diff
@@ -122,6 +122,211 @@ version(mir_bignum_test)
+    // No liability is assumed for incidental or consequential damages in connection with or arising out of the use of the information or programs contained herein. 
+    import mir.bitop: ctlz;
+    import mir.utility: _expect;
+
+    enum shift = (W.sizeof * 8);
+    enum b = ulong(1) << shift;
+
+    // Convert all of our input arguments to little-endian
+    auto u = _u.leastSignificantFirst();
+    auto v = _v.leastSignificantFirst();
+    auto q = _q.leastSignificantFirst();
+    auto r = _r.leastSignificantFirst();
+
+    assert(m != 0 && n != 0);
+    assert(_u.coefficients.length >= m + 1, "Dividend array is too small (should be m+1)");
+    assert(_v.coefficients.length >= n, "Divisor array is too small");
+    assert(_r.coefficients.length >= n, "Remainder array is too small");
+    assert(v[n - 1] != 0);
+
+    // Handle a common edge case with division (where the dividend is smaller then the divisor)
+    // We don't care about how big the quotient array is here (and in fact, our math will fail)
+    // since we just dump the dividend into the remainder 
+    if (_expect(m < n, false))
+    {
+        foreach(i; 0 .. m)
+        {
+            r[i] = u[i];
+        }
+        return;
+    }
+
+    // From this point on, now we care if the quotient array is appropriately sized
+    assert(_q.coefficients.length >= (m - n) + 1, "Quotient array is too small");
+
+    // According to Knuth, when the divisor length is equal to one,
+    // we should fall back to a simplified algorithm
+    if (_expect(n == 1, false))
+    {
+        uint k = 0;
+        foreach_reverse(j; 0 .. m)
+        {
+            // Same optimization as down below (left-shifting / ORing in place of multiplication / addition)
+            q[j] = cast(uint)(((cast(ulong)(k) << shift) | u[j]) / v[0]);
+            k = cast(uint)(((cast(ulong)(k) << shift) | u[j]) % v[0]);
+        }
+        _r.leastSignificant = k;
+    }
+    else
+    {
+        // pre-condition for this algorithm to succeed.
+        // assert(m >= n, "failed algorithm pre-condition");
+
+        // D1: Normalize
+        // In TAOCP, this is specified as b/(v[n - 1] + 1)
+        auto s = cast(short)ctlz(v[n - 1]);
+        assert(s <= 31 && s >= 0);
+
+        // This normally uses a for loop, but we just use
+        // a LLV primitive here instead.
+        _v.smallLeftShiftInPlace(s);
+        _u.smallLeftShiftInPlace(s);
+
+        foreach_reverse(j; 0 .. (m - n) + 1)
+        {
+            // D3: Calculate qhat, rhat == remainder
+            // In TOACP, this is specified as u[j+n]*b + u[j+n-1] / v[n - 1]
+            // but, since we assume b is a power of 2, we can rewrite it to just
+            // use left-shifting (as left-shifting effectively multiples by powers of 2)
+            // and use an OR in place of the addition (as we've cleared the bottom 32 bits)
+            ulong qhat = ((cast(ulong)(u[j+n]) << shift) | u[j+n-1]) / v[n-1];
```

Comment:
add temporal like `auto ut = (ulong(u[j + n]) << shift) | u[j + n - 1];` and use it below. us